### ISSUE TITLE
Community Vote: UI / JavaScript Framework

### ISSUE THREAD

**Issue by AzorianSolutions** [PowerDNS-Admin/PowerDNS-Admin#1069]
I was thinking it would be nice to give the template and JavaScript organization a healthy overhaul. It's getting fairly cluttered and somewhat difficult to navigate.

I see a great scenario to drop in a basic reactive UI framework. This would inherently provide better template organization while properly decoupling the HTML and JavaScript code. This would also pave the road for a real time UI implementation.

I could also work in better deep linking for things like tabbed views. This would be nice for easier refreshing.

I'm looking for feedback on whether or not this type of overhaul would be accepted or rejected. Thoughts or concerns?

**Comment by zoeller-freinet** [comment]
With [AdminLTE](https://adminlte.io/) (which itself builds upon bootstrap), PowerDNS-Admin already uses such a UI framework. PDA's front-end code is indeed unnecessarily unwieldy and disorganized in many places, and a thorough refactor is indeed due (PRs welcome!). I think it has more to do with how PowerDNS-Admin makes use of its dependencies than what dependencies it uses - AdminLTE and tidy & modern UI code are not mutually exclusive and should entirely be in the realm of possibility.

Incidentally, I have been mulling over how going into a feature freeze period sometime soon, focusing exclusively on bugfixes, dependency updates and general refactoring, could do a lot of good for PowerDNS-Admin.

**Comment by AzorianSolutions** [comment]
@zoeller-freinet I don't believe we are on the same page with this. I was not referring to CSS / HTML UI frameworks such as Bootstrap. I was referring to JavaScript frameworks which incidentally provide some UI code management. Think VueJS, React, KnockoutJS, etc...

I don't know the history of progress and haven't looked so I can't add my two cents on the value of a feature freeze. What I can tell you from years of experience, is that for any project that has growing demands, a well-organized and thought out foundation is always the key to proper growth. Hence my desire to contribute some work toward that foundation.

Thoughts?

**Comment by zoeller-freinet** [comment]
Thank you for elaborating, it's a lot clearer to me now what you're after. It should probably be done at some point. No solid concerns spring to mind currently, other than it maybe not having highest priority right now. As for choice of JS framework, do you have a preference?

Other thoughts / concerns?

**Comment by jbe-dw** [comment]
Hello,

I can't really add a valued answer on the original question as I am certainly not a dev and have zero skills about JS technologies. I just get in this talk to confirm the need to refactor the app. 

It's functional but a pain to patch, and the codebase is not homogeneous at all. I have used some time to make patches and it's now close to what I need (I could just send some eye blinks to @RGanor about personal apikeys one more time). But spending time on it clearly shows that it lacks structured design and application wide consistency (the API is functionnal but it clearly needs to be rewritten).

From the JS perspective, as a user with no knowledge about deployments, I would expect that installing the application involves no manual action. It sometimes breaks because yarn repositories may not be available or it had a broken dependency recently.

@zoeller-freinet Before you kicked in, the management of this project were erratic, so having a freeze is not a problem for me, if you give WIP status updates frequently.

(a lot of work is needed to cleanup the github space first: close 3y old bug or request or dead PRs)

**Comment by AzorianSolutions** [comment]
@zoeller-freinet You're welcome. First, to answer your question, I'm thinking VueJS myself. I came to this based on active maintenance commitment, ease of use, and application of accepted and personal standards. To me, this particular choice sees like a fairly wise and safe investment for the (foreseeable) future.

As for your concern about priorities, I can certainly appreciate that. Here is what I can offer; given that I am not a solid Python dev, I have a slight learning curve to endure as I continue to really sink my teeth in. So I can certainly manage making some smaller modifications from time to time and occasionally overseeing other changes to a degree. As far as what really motivates me though are two key elements;

1) Projects that provide added value to my organizations.
2) Projects that make great use of my skill sets which inherently drives higher productivity.

I think from an engineering productivity standpoint, perhaps you should consider the following;

1) An important factor to drive development participation is barrier to entry (or ideally the lack thereof). This is why I also committed to cleaning up the broken Docker implementation that is quite stale. I think it is true of many devs to be less motivated to work on something if it takes a lot of time and effort to even get started properly.

2) If you do accept that this type of change is inevitable, then it is important to think through exactly how this change fundamentally affects the implementation of features. For example, say you have an open FR currently that requests a new button control for a function. The way you implement that feature today is completely different from how you would implement that same feature post-framework implementation. For this reason, it's important to consider how much that can double up on work as a direct function of time from implementation.

I think some areas of this app could use some new life pumped into it and I'm one of the few currently interested in trying. Run with the opportunity or no? I think it's definitely a potential and likely pain point during transition but certainly a band-aid that needs ripped off so to speak.

**Comment by AzorianSolutions** [comment]
@jbe-dw I agree with you. I will say this app is not in terrible condition by any means. I have dealt with a wide variety in my time so far and this one still has me interested after digging in That's always a great sign IMHO.

Pro-tip, if you understand good engineering practices in one form of programming, then you likely understand many of the same in JS. Of course, there are many differences as well such as the very asynchronous nature of apps but many common principals still apply. Decoupling application layers still makes for productive maintenance and provides clear paths to code organization.

**Comment by AzorianSolutions** [comment]
@jbe-dw I appreciate that :-)

Forgive me if some of this is obvious to you, I'm just having a hard time gauging what you are and aren't comfortable with the concepts of. Frameworks are a key word because they are in ways much like a spoken language or a company handbook. They provide a universal set of practices and rules for how something should be done. Without such guides, a code base can quickly become the wild west. This is only quantified by smaller contributions that will often replicate the running pseudo standard of implementation, quickly scaling the problem larger.

For the record, I'm not advocating tearing out the Flask / SQLAlchemy foundations right now. My solution can integrate on top of them as they are which would take the app in a more modern direction. This would essentially work towards decoupling the UI from the back end further. The final result would be a back-end that just serves static assets and an HTTP API which the UI would consume. This is often considered most productive because it forces you to consume your own API actively with everything you do. If you can imagine, this can result in fewer API development issues and better testing.

Simply put, with the modern app today, there should be a clear division between your front-end and back-end services.

Hope this helps!

**Comment by jbe-dw** [comment]
Sorry I was talking about "modern JS frameworks". I'm a senior Sysops doing automation (meaning: coding everything I build).
About flask it is an example as well. As i'm not in JS, I gave Python examples.

This helps, and as I mainly consume the API, the idea of a frontend that do the same looks very good to me.

**Comment by zoeller-freinet** [comment]
@AzorianSolutions I wouldn't go so far as to say that the introduction of a JS Framework like VueJS is inevitable, but do count me among the advocates. Further, I fully agree that maintaining a low barrier to entry for developers is most often crucial for keeping projects healthy, it certainly applies to PDA. Incorporating a JS framework would, on its own, increase barrier to entry, as you'd then additionally need to know how to use that framework, at least when working on the front-end. But consistently applied, that should amortize in the long run as a result of the architectural improvements.

I appreciate that you're willing to take a step forward here. The realization of the idea does not stand or fall with me - we could wait for more opinions, otherwise feel free to start working on it anytime of course. We could maintain a separate branch with feature parity to `master`, into which PRs are continuously merged to reorganize the app. Once it's mature, it could be decided that it will be merged into `master`. As opposed to continuously updating `master`, this would have the advantage of structuring the work better and keeping the look and feel of `master` consistent until merge. I expect it could take more than, say a few weeks until the overhaul nears the finish line and changes settle down, so I also expect that new versions of PDA will or at least should be released during that time.

@jbe-dw 
> Before you kicked in, the management of this project were erratic, so having a freeze is not a problem for me, if you give WIP status updates frequently.

Thanks! I feel you're making me a bit bigger than I am here. I'm just somebody who likes to help with keeping PDA in good shape - me and my organization depend on it - while having fun doing so, with the very limited resources that I have available for it. Currently I can't realistically put more than 2-3 hours per week into PDA in a sustained way. That's not necessarily to say that I'd be unable to manage a temporary feature-freeze while giving status updates, but just to give you an idea. Happy if my contributions matter :)

**Comment by AzorianSolutions** [comment]
@zoeller-freinet is English your first language? Either way, you're very well spoken! I appreciate your view and completely agree. I also recognize the fallacy in the proposal. Specifically, the part about a new framework increasing the barrier to entry. Just as you said, I believe that will be mitigated by the overall organizational gains but it does certainly add to it nonetheless.

I think a good median could be achieved here so I will take you up on that advice. I will start a new branch to hold the work of these efforts. Thanks @zoeller-freinet!

**Comment by JCCouture1** [comment]
Hello everyone 
Please how do I assign the prev and next on music button ?
